e2e-tests/src/test/adder.rs

```diff
@@ -13,14 +18,31 @@ pub async fn adder() -> Result<()> {
     let config = setup_test();
 
     let (conn, _authority, account) = basic_test_context(config).await?;
-    let contract = AdderInstance::new(
+
+    let contract = Arc::new(AdderInstance::new(
         &config.test_case_params.adder,
         &config.test_case_params.adder_metadata,
-    )?;
+    )?);
+
+    let listen_conn = conn.clone();
+    let listen_contract = contract.clone();
+    let (tx, mut rx) = unbounded();
+    let listen = || async move {
+        listen_contract_events(&listen_conn, &[listen_contract.as_ref().into()], tx).await?;
+        <Result<(), anyhow::Error>>::Ok(())
+    };
+    let join = tokio::spawn(listen());
 
     let increment = 10;
     let before = contract.get(&conn).await?;
+
     contract.add(&account.sign(&conn), increment).await?;
+
+    let event = rx.next().await.context("No event received")??;
```

Comment:
Will `next()` resolve immediately? i.e. it will either return or not, without waiting. So we're assuming here that by the time we get to this line, the `contract.add(...)` transaction has already been processed by the chain?

Reply:
I'm not sure what the question is... `next` returns a future, it will not resolve immediately. Even if technically the event already occurred, there is no guarantee on how long it's going to take to fetch that event into the client. There is indeed a way to see events "immediately" after a contract call - a call result can return all the events in the block that the call ended up in. However, I intentionally didn't go for that method, because I think it's less flexible - it only allows you to observe events caused by your own calls. If anything, I'd say that can be a secondary feature, but this listening is something that must be there.

Reply:
> There is indeed a way to see events "immediately" after a contract call - a call result can return all the events in the block that the call ended up in. However, I intentionally didn't go for that method, because I think it's less flexible - it only allows you to _observe events caused by **your own calls**_.

I think that's exactly what we're interested in here, aren't we? Otherwise we might get here an event that was created by someone else's call that happened right after we sent the txn but wasn't result of that txn.

Reply:
> I think that's exactly what we're interested in here, aren't we?

Depends on what you mean by "interested". If by "interested" you mean for the purposes of these tests, then yes. But that's not the (whole) point of the exercise. I would say the more important goal is to provide tooling for people to interact with the contracts, and subscribing to events without doing calls is the more flexible approach for that.

Reply:
OK, I'm convinced. Thanks.